### PR TITLE
Create `unnecessary_send_constraint` lint for `&(dyn ... + Send)`

### PR DESCRIPTION
**Pull request by john-h-k**
Fixes #110937

### CODE REVIEW

File: library/core/src/cell.rs
Comment (john-h-k) [inline library/core/src/cell.rs:2249]: This line currently causes compilation to fail as it is unrecognised lint

**Comment by john-h-k** [comment]
Also, it appears no lint is generated for duplicate auto traits (like `&(dyn Any + Any + Send)`). Is there a reason for this? If not I can add that to as the logic is similar ish (in a separate PR)

File: library/core/src/cell.rs
Comment (Urgau) [reply library/core/src/cell.rs:2249]: Gating the attribute on stage>=1 with `bootstrap` cfg should do it:
```suggestion
#[cfg_attr(not(bootstrap), allow(useless_send_constraint))]
```

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (est31) [inline compiler/rustc_lint/src/useless_send_constraint.rs:14]: This statement is wrong. If the base type implements [`Sync`](https://doc.rust-lang.org/stable/std/marker/trait.Sync.html) then the reference *can* be sent across threads.

The issue is that references of `dyn (Send + ...)` can't be sent across threads, at least as long as `...` doesn't contain a `Sync` bound.

Comment (john-h-k) [reply compiler/rustc_lint/src/useless_send_constraint.rs:14]: Thanks have addressed

Comment (john-h-k) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: Just to clarify the behaviour:

* `&dyn Send` should be changed to `&dyn Any`
* `&dyn ... + Send` should have send removed, unless sync is specified
* Mutable references should be ignored

Have only been learning rust for a couple of weeks so my understanding of the threading model is a bit weak, but I will read more about it to try and improve my understanding

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (est31) [inline compiler/rustc_lint/src/useless_send_constraint.rs:20]:
```suggestion
    /// References cannot be sent across threads unless they have a `Sync` bound, so constraining them to `Send` without `Sync` is useless.
```

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (est31) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: > Have only been learning rust for a couple of weeks

Don't worry, we are all learners!

The `Send` bound is never useful. If `Sync` is not specified, adding `Send` won't make the reference `Send`, it will always not be `Send`. If `Sync` is specified, `Send` is always implemented whether it's specified or not.

I'm not sure actually what's going on with mutable references, what would be your reasoning why they should be ignored?

Comment (john-h-k) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: > The Send bound is never useful. If Sync is not specified, adding Send won't make the reference Send, it will always not be Send. If Sync is specified, Send is always implemented whether it's specified or not.

Ah right that makes sense, like with copy and clone. When it's `Sync + Send`, I'll give a different message that it's redundant due to sync being a super trait

> I'm not sure actually what's going on with mutable references, what would be your reasoning why they should be ignored?

I don't think there was any real logic behind it, I think I assumed based off of the original issue + the fact it seemed to break a lot of code when I removed a `+ Send` on a `*mut (dyn Any + Send)`, whereas when I removed it on non-mut references it didn't break anything, but that must've just been a coincidence...

**Comment by john-h-k** [comment]
Is there any documentation on the `.ftl` files and how the DSL for them works? I'm struggling a bit with getting sub diagnostics to work

**Comment by est31** [comment]
@john-h-k you can take a look at the first two sub-chapters 50.1 and 50.2 of the [diagnostics chapter](https://rustc-dev-guide.rust-lang.org/diagnostics.html) in the rustc book, or alternatively at some of the open PRs to migrate diagnostics.

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (lukas-code) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: > I'm not sure actually what's going on with mutable references

`&mut T` is `Send` iff `T` is `Send`, even though this is kind of badly documented. [(proof)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=190547f8744598ca50fd19ba8aa3654c)

So this lint should only trigger for exactly `&dyn Send`, not `&mut dyn Send`.

Comment (m-ou-se) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: To clarify: in `&dyn X + Send`, neither the `X` nor the `Send` is about the reference. It means that the thing that the reference points at implements both `X` and `Send`. But whether or not something implements `Send` isn't relevant if you only have a reference to it, because you can't send it if you don't own it.

Comment (est31) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: I've made a PR to add it to the docs: #111065

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (lcnr) [inline compiler/rustc_lint/src/useless_send_constraint.rs:21]: nit wrt naming:

I prefer "unnecessary" over "useless" here. useless feels a lot more negative to me and even sounds like it kind of blames the author for writing bad code :thinking:

Comment (john-h-k) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: done

**Comment by john-h-k** [comment]
> Also, `tests/ui/lint` tests seem to have really inconsistent naming - when should I use hyphens vs underscores in file names?

bumping this as the tests are the only bit left really

**Comment by oli-obk** [comment]
Nominating for T-lang discussion about whether we want this new lint in the compiler enabled by default. The reasoning is explained in https://github.com/rust-lang/rust/issues/110937, but the TLDR is that a `Send` bound on a trait object behind an immutable reference is not changing anything, so it can be removed

File: compiler/rustc_lint/src/useless_send_constraint.rs
Comment (lcnr) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: please also update the PR name to reflect this change

Comment (john-h-k) [reply compiler/rustc_lint/src/useless_send_constraint.rs]: done

**Comment by nikomatsakis** [comment]
What happens if you get a compilation error? Ideally we would suggest replacing `&dyn Any + Send` with `&dyn Any + Send + Sync` in that case.

**Comment by nikomatsakis** [comment]
Discussed in the @rust-lang/lang meeting -- @m-ou-se can you say a bit more about the motivation here? Is this an issue that you've seen cause confusion in practice?

**Comment by m-ou-se** [comment]
@nikomatsakis The problem is not confusion (with Sync or w/e). The problem is accidentally making an API too strict.

We accidentally made std's PanicInfo::payload() return a &(dyn Any + Send) instead of just a dyn Any (because it usually refers to the contents of a Box<dyn Any + Send>), which means we now cannot safely construct a PanicInfo for a non-Send payload because we promised it to be Send, even though that bound is useless.

Removing the Send trait from the signature breaks a few crates, because they pass it to functions that expect a &(dyn Any + Send). Those functions should've just taken a &dyn Any, but the type was simply coped from the payload() return type (which also should've been &dyn Any).

**Review by davidtwco — APPROVED** [review]
Implementation of the lint looks good to me but I'm not clear whether it was decided this was desirable, so I'll leave that to others in this thread.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, and the motivation makes sense: this is generally redundant and including it may make an API stricter than desired.

@rfcbot merge

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (RalfJung) [inline compiler/rustc_lint/src/unnecessary_send_constraint.rs:20]: *Shared* references, specifically

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (RalfJung) [inline compiler/rustc_lint/src/unnecessary_send_constraint.rs:23]: Shouldn't the recommendation be replace `Send` by `Sync`? That seems more likely to be what the author intended.

File: compiler/rustc_lint/messages.ftl
Comment (RalfJung) [inline compiler/rustc_lint/messages.ftl:197]: Replacing by `Any` is a very strange suggestion. Why would `type_id` machinery be necessary? Can't the `'static` bound cause a problem?

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (m-ou-se) [reply compiler/rustc_lint/src/unnecessary_send_constraint.rs:23]: See this comment: https://github.com/rust-lang/rust/pull/110961#issuecomment-1550271436

In those situations, it stems from a `Box<dyn Any + Send>` to which a &dyn was made. It's easy to just copy `Any + Send`, but the Send part should be removed for shared references.

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (m-ou-se) [inline compiler/rustc_lint/src/unnecessary_send_constraint.rs:20]: This explanation seems wrong/unrelated. This sentence talks about the reference itself being `Sync`/`Send`, but the `Any` and `Send` in `&(dyn Any + Send)` is about the object the reference points to, not about the reference itself. `Send` is useless there because a `&T` does not allow sending the `T`.

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (RalfJung) [reply compiler/rustc_lint/src/unnecessary_send_constraint.rs:23]: That seems like a fairly special case to me, I would not expect this to be common enough to warrant a general recommendation of dropping the `Send`. The more symmetric thing to do would be to have `&dyn Any+Sync`, to get a shared reference that can still be shared across thread boundaries. Of course then you'd need `Box<dyn Any+Send+Sync>` for the owned case.

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (RalfJung) [reply compiler/rustc_lint/src/unnecessary_send_constraint.rs:20]: I think it talks about the object being `Sync`, which is exactly what is needed to make the reference `Send`, which presumably is what the programmer intended to achieve.

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (oxalica) [inline compiler/rustc_lint/src/unnecessary_send_constraint.rs:13]: This asserts compile_fail for a typo, while the lint itself is only WARN. Do you mean
```suggestion
    /// #[deny(unnecessary_send_constraint)]
    /// fn foo(_: &(dyn Any + Send)) {}
```
?

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (oxalica) [inline compiler/rustc_lint/src/unnecessary_send_constraint.rs:8]:
```suggestion
    /// The `unnecessary_send_constraint` lints unnecessary constraint of references to `Send`.
```

**Comment by nikomatsakis** [comment]
It looks like we agreed to this. I'm going to drop nomination.

@rustbot labels -I-lang-nominated

**Comment by nikomatsakis** [comment]
@rustbot author

The team role is done here, FCP is complete, we are now waiting on the author to address feedback though.

File: compiler/rustc_lint/messages.ftl
Comment (john-h-k) [reply compiler/rustc_lint/messages.ftl:197]: I wasn't sure what else to put, this is for the scenario where `Send` is the only trait specified (so removing it would be invalid)

File: compiler/rustc_lint/src/unnecessary_send_constraint.rs
Comment (john-h-k) [reply compiler/rustc_lint/src/unnecessary_send_constraint.rs:23]: So it should recommend replacing `Send` with `Sync` instead? just confirming before making the change

**Comment by john-h-k** [comment]
@rustbot ready

(still 1 part of the review I am unsure about so have asked for clarification, but other comments have been addressed)

File: library/std/src/io/error.rs
Comment (steffahn) [inline library/std/src/io/error.rs:698]: `dyn Error + Send + Sync` has a large API surface that `dyn Error + Sync` doesn’t have.

Comment (steffahn) [reply library/std/src/io/error.rs:698]: Also, in any case, changing these methods is a breaking change the impact of which I’m not sure has been tested anywhere yet.

Comment (oxalica) [reply library/std/src/io/error.rs:698]: > that `dyn Error + Sync` doesn’t have. Like an [`is` method](https://doc.rust-lang.org/std/error/trait.Error.html#method.is-2) and a [`downcast_ref`](https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-2) method.

[`dyn Error`  also implements](https://doc.rust-lang.org/std/error/trait.Error.html#impl-dyn+Error) `is` and `downcast_*` with exact the same signature.

Comment (oxalica) [reply library/std/src/io/error.rs:698]: > [`dyn Error` also implements](https://doc.rust-lang.org/std/error/trait.Error.html#impl-dyn+Error) `is` and `downcast_*` with exact the same signature.

Oops, my bad. We cannot use `dyn Error`'s methods on `dyn Error + Sync`, though it seems odd to me.

**Comment by Dylan-DPC** [comment]
@john-h-k if you can rebase and resolve the conflicts, we can push this for a review

**Comment by alex-semenyuk** [comment]
@john-h-k
From wg-triage. Closed this PR due to inactivity. Feel free to reopen or raised new one. Thanks for your efforts.

**Pull request closed without merging**
